runtime/src/in_mem_accounts_index.rs

```diff
@@ -465,46 +467,83 @@ impl<T: IndexValue> InMemAccountsIndex<T> {
+    /// Note that even if entry DID exist at 'other_slot', the above conditions apply.
     fn update_slot_list(
         slot_list: &mut SlotList<T>,
         slot: Slot,
         account_info: T,
-        _other_slot: Option<Slot>,
+        mut other_slot: Option<Slot>,
         reclaims: &mut SlotList<T>,
         previous_slot_entry_was_cached: bool,
     ) -> bool {
         let mut addref = !account_info.is_cached();
 
-        // find other dirty entries from the same slot
-        for list_index in 0..slot_list.len() {
-            let (s, previous_update_value) = &slot_list[list_index];
-            if *s == slot {
-                let previous_was_cached = previous_update_value.is_cached();
-                addref = addref && previous_was_cached;
-
-                let mut new_item = (slot, account_info);
-                std::mem::swap(&mut new_item, &mut slot_list[list_index]);
-                if previous_slot_entry_was_cached {
-                    assert!(previous_was_cached);
-                } else {
-                    reclaims.push(new_item);
-                }
-                slot_list[(list_index + 1)..]
-                    .iter()
-                    .for_each(|item| assert!(item.0 != slot));
-                return addref;
-            }
+        if other_slot == Some(slot) {
+            other_slot = None; // redundant info, so ignore
         }
 
-        // if we make it here, we did not find the slot in the list
-        slot_list.push((slot, account_info));
+        // There may be 0..=2 dirty accounts found (one at 'slot' and one at 'other_slot')
+        // that are already in the slot list.  Since the first one found will be swapped with the
+        // new account, if a second one is found, we cannot swap again. Instead, just remove it.
+        let mut found_slot = false;
+        let mut found_other_slot = false;
+        (0..slot_list.len())
+            .into_iter()
+            .rev() // rev since we delete from the list in some cases
+            .for_each(|slot_list_index| {
+                let (cur_slot, cur_account_info) = &slot_list[slot_list_index];
+                let matched_slot = *cur_slot == slot;
+                if matched_slot || Some(*cur_slot) == other_slot {
+                    // make sure neither 'slot' nor 'other_slot' are in the slot list more than once
+                    let matched_other_slot = !matched_slot;
+                    assert!(
+                        !(found_slot && matched_slot || matched_other_slot && found_other_slot),
+                        "{:?}, slot: {}, other_slot: {:?}",
+                        slot_list,
+                        slot,
+                        other_slot
+                    );
+
+                    let is_cur_account_cached = cur_account_info.is_cached();
+                    let mut new_item = (slot, account_info);
```

Comment:
nit: This `new_item` can be moved inside the `if` block just below, yes?
https://github.com/solana-labs/solana/blob/8c08a0940ad28512c224fa931d8996983a57496c/runtime/src/in_mem_accounts_index.rs#L517-L521